README.md
# jobsworth-installer
[Jobsworth](https://github.com/ari/jobsworth) installer and migrator from hosted [ClockingIT](http://www.clockingit.com/)

This project has the purpose of helping you install [jobsworth](https://github.com/ari/jobsworth) in an Ubuntu and Ubuntu Server OS, and also to help you migrate from the hosted verion of [ClockingIT](http://www.clockingit.com/)

At the moment the instructions are very raw, and the migration script might be unstable so be careful using it. You can choos to make an automatic install with the provided scripts or a Manual install following the steps detailed below.

## Automatic Installation of Jobsworth
You can install Jobsworth and also migrate your data if you wish, by using the scripts provided in this project.
The scripts will try to download and install everything you need.

### Instructions
To download the script and execute all you have to do is execute the following commands in the linux console:
```sh
wget https://github.com/frank-orellana/jobsworth-installer/archive/v0.3.1.tar.gz
tar xzf v0.3.1.tar.gz
cd jobsworth-installer-0.3.1
chmod +x install-jobsworth.sh
sudo ./install-jobsworth.sh
```

Follow the steps on screen and when everything is finished without errors you will be able to access your brand new installation of jobsworth in http://your-ip:8080/ with the user: `admin` and password: `password`. It might take a while the first time.

<details>
  <summary>Detailed instructions and details for automatic Install</summary>

All this scripts have been tested ONLY on new installations of Debian 8.8, Ubuntu 15.10, 16.04 and 16.10

There's also a VirtualBox VM ready to try if you wish to download it from [Dropbox](http://bit.ly/2niDqXL) with the instructions to use it [here](https://github.com/frank-orellana/jobsworth/releases): 

### Detailed instructions and details for automatic Install And/Or Data migration
1. Download the full project
1. Extract the files to a folder IN your server
1. Move to the `jobsworth-installer` directory
1. Make the file executable `chmod +x install-jobsworth.sh`
1. For **Installation** execute the file: `sudo ./install-jobsworth.sh`. 

The script will **automatically**:
 1. Download and install MySql (It will ask you to set the root password)
 1. Create a new database named `jobsworth` and a database user named `jw`
 1. Download, Install and configure Java JDK and Tomcat9
    1. You will be able to access tomcat welcome page at http://your-ip:8080/, and also the managemente addreses in http://your-ip:8080/host-manager and http://your-ip:8080/host-manager with the user and password `jobsworth` from your os and local network only.
        2. The installation path will be /opt/tomcat. For more details see the script
 1. Download, install and configure Jobsworth
 1. After the installation is complete you can access jobsworth through http://your-ip:8080/ with the user: `admin` and password: `password`. It might take a while the first time
1. For **Data Migration** view the instructions below
 
 To see the full details see the [Migration Guide](https://github.com/frank-orellana/jobsworth-installer/tree/master/migration-from-hosted)

The scripts need to be executable. If they are not you will have to execute this before trying to execute them:
```sh
chmod +x install-jobsworth.sh
chmod +x migration-from-hosted/migrate-from-hosted.sh
```

### Parameters of the script:
You can prevent parts of the script from executing like this:
`sudo NOXX=1 NOYY=1 ./install-jobsworth.sh`

The options you can prevent are:
<table>
<tr><th>OPTION</th><th>What it prevents</th></tr>
<tr><td>NO_UPD_REP</td><td>The update of repositories (apt-get update) for example for when you are running the script a second time</td></tr>
<tr><td>NO_DB_SERVER</td><td>The installation of mariadb-server</td></tr>
<tr><td>NO_DB</td><td>The creation of the jobsworth database</td></tr>
<tr><td>NO_JDK</td><td>The installation of openjdk-8-jdk</td></tr>
<tr><td>NO_TOMCAT</td><td>Tomcat download and installation</td></tr>
</table>

</details>

## Manual Installation of Jobsworth
If you do not wish to use the automatic installation script, there is also a tutorial to install jobsworth from scratch in a clean Ubuntu 16.10 Server & Desktop distro.

The full steps can be found in [here](https://github.com/frank-orellana/jobsworth-installer/wiki/Manual-installation).
<details>
<summary>Click here to view the steps required to manually install jobsworth</summary>

### Overview of the steps
Basically what you will do is:
1. Install Ubuntu 15.10 or above
2. Install and configure MariaDB (MySql)
3. Install and configure Java JDK 8
4. Install and configure Tomcat
5. Install and configure the Jobsworth WAR file

Full steps [here](https://github.com/frank-orellana/jobsworth-installer/wiki/Manual-installation).

Also there is a VM already configured if you prefer. You can download it from [dropbox](http://bit.ly/2niDqXL).
</details>

## Migration from Hosted ClockingIT
With these steps you should be able to migrate your projects, tasks, users and most of the data from the hosted version of ClockingIT to your installation of jobsworth. You need to have already installed jobsworth for this to work.
### Automatic migration steps
For the migration you will need to:
1. Download the full project, extract it in a folder IN your server
1. Go to the folder `jobsworth-installer/migration-from-hosted`
1. Download the sql dump from the hosted version of ClockingIT
1. Make the script executable `chmod +x migrate-from-hosted.sh`
1. Run the script passing as a parameter the dump file name: <br>  `./migrate-from-hosted.sh my_dump_file.sql`

The full steps for the migration can be found [here](https://github.com/frank-orellana/jobsworth-installer/tree/master/migration-from-hosted)


----

[![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.me/FranklinOrellana)


changelog.md
v0.3.1
-Small tutorial and logging fixes

v0.3.0
-Updated and simplified installation instructions
-Added support for debian 8.8
-Improved tomcat service memory configuration depending on amount of memory available
-Improved jdk detection and installation
-Fixed problem when a jdk is already installed
-Fixed tomcat download link
-Fixed & improved installer logging
-Added script parameters so some parts are not executed
-Tomcat & jobsworth binaries are now downloaded into the same folder as the installer
-Support continue for failed downloads of Tomcat & jobsworth


Tested with:
Ubuntu: Server 16.10, Debian: 8.8
Jobsworth: 5.0 beta 2
MySql: mariadb-server 10.0.29
Java JDK: java-1.8.0-openjdk-amd64
Tomcat: v8.5.15


v0.2.0

Changes
-Added support for ubuntu 15.04
-Changed db engine from mysql to mariadb (be careful if you already use mysql you can choose not to install it)
-Instead of default jdk, the installer now uses openJDK 8
-Tomcat version used is now the stable one (8.5.14)
-Fixed & improved installer logging
-improved tomcat service memory configuration

Tested with:
Ubuntu: Desktop 16.10, Server 16.10, Server 15.04
Jobsworth: 5.0 beta 2
MySql: mariadb-server 10.0.25
Java JDK: java-1.8.0-openjdk-amd64
Tomcat: v8.5.14



v0.1.0
This installer was tested on:
Ubuntu: Desktop 16.10, Server 16.10
Jobsworth: 5.0 beta 2
MySql: mysql-server 5.7.17
Java JDK: java-1.8.0-openjdk-amd64
Tomcat: v9.0.0.M19


install-jobsworth-from-scratch.sh
#!/bin/bash

#This file is part of the jobsworth-installer program
#https://github.com/frank-orellana/jobsworth-installer
#Copyright (C) 2017  Franklin Orellana
#This program is free software; you can redistribute it and/or modify
#it under the terms of the GNU General Public License as published by
#the Free Software Foundation; either version 3 of the License, or
#(at your option) any later version.
#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.
#You should have received a copy of the GNU General Public License
#along with this program; if not, write to the Free Software Foundation,
#Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
#WARNING:
#-This scrpit will overwrite info you may have on the jobsworth

echo "Starting installation..." 

pushd `dirname $0` > /dev/null
mydir=`pwd`
popd > /dev/null

logfile="$mydir/install.log"
echo "* logfile: $logfile"

if (( $EUID != 0 )); then
    echo 'ERROR: You dont have sudo rights. Execute the file like this: sudo install-jobsworth-from-scratch.sh' | tee $logfile
    exit
else
	echo '* Starting process...' | tee $logfile ;
fi

echo "* Updating repositories..."
apt-get -qq update

echo "* Installing mysql (MariaDB)" | tee -a $logfile

echo "*   If it is not already installed it will ask you for a new password for "
echo "*   the ROOT user. DON'T FORGET THAT PASSWORD!!!"
echo "*   It will be asked later to create the default database"
echo "* Downloading..."
apt-get -qqy install mariadb-server

echo "* Creating Jobsworth database."
echo "* Executing create-database.sql script... Enter mysql (MariaDB) root password:" | tee -a $logfile
mysql -u root -p < installer-resources/create-database.sql
echo
echo "******************************************************************************"
echo "* Jobsworth database created. To access it you can type 'mysql -u jobsworth -p'" | tee -a $logfile
echo "******************************************************************************"
echo
echo "* Installing JAVA" | tee -a $logfile

apt-get -qqy install openjdk-8-jdk  | tee -a $logfile

JHF=$(update-java-alternatives -l | grep -o '/.*jvm/java-.*$')/jre
if [ -z "$JHF" ]; then echo "* ERROR: JAVA_HOME Not Found. Exiting" | tee -a $logfile; exit; else echo "* JAVA_HOME FOUND: $JHF" | tee -a $logfile; fi

groupadd tomcat
useradd -s /bin/false -g tomcat -d /opt/tomcat tomcat

cd /tmp

echo "* Downloading TOMCAT..." | tee -a $logfile
wget -nv -nc http://www-us.apache.org/dist/tomcat/tomcat-8/v8.5.14/bin/apache-tomcat-8.5.14.tar.gz | tee -a $logfile

echo "* TOMCAT dowloaded" | tee -a $logfile

mkdir /opt/tomcat
tar -xzf apache-tomcat-8*tar.gz -C /opt/tomcat --strip-components=1

echo "* TOMCAT extracted" | tee -a $logfile

echo "* Configuring TOMCAT..." | tee -a $logfile
cd /opt/tomcat
chgrp -R tomcat /opt/tomcat
chmod -R g+r conf
chmod g+x conf
chown -R tomcat webapps/ work/ temp/ logs/
chmod 775 -R conf/ logs/ webapps/

cd conf
find . -type f -exec chmod 664 -- {} +

cd ../webapps
find . -type f -exec chmod 664 -- {} +

cd $mydir

chmod +x installer-resources/configure-tomcat.sh
installer-resources/configure-tomcat.sh


echo "* TOMCAT has been installed. Please check its installation visiting http://localhost:8008 OR HTTP://<THIS_COMPUTER_IP>:8080 before continuing"
read -n 1 -s -p "Press ENTER to continue"
echo
echo "* Stopping TOMCAT"
systemctl stop tomcat

echo "* Downloading and Installing Jobsworth" | tee -a $logfile
cd /var
mkdir -p jobsworth/assets

chown -R tomcat jobsworth
chgrp -R tomcat jobsworth
chmod -R 775 jobsworth

cd /opt/tomcat
rm -rf webapps/ROOT

cd /opt/tomcat/webapps
echo "* Downloading WAR" | tee -a $logfile
wget -nv -nc https://github.com/ari/jobsworth/releases/download/5.0b2/ROOT.war  | tee -a $logfile
echo "* WAR Downloaded" | tee -a $logfile
chown tomcat ROOT.war
chgrp tomcat ROOT.war

systemctl start tomcat
echo "* TOMCAT STATUS:" | tee -a $logfile
systemctl status tomcat | grep Active: | tee -a $logfile

chmod 664 /opt/tomcat/logs/*
echo
echo "******************************************************************************"
echo "******************************************************************************"
echo "JOBSWORTH HAS BEEN INSTALLED"  | tee -a $logfile
echo "THE INSTALLATION IS COMPLETE TYPE HTTP://LOCALHOST:8080 in the browser from "
echo "this computer OR HTTP://<THIS_COMPUTER_IP>:8080 FROM ANY OTHER"
echo "******************************************************************************"
echo "******************************************************************************"
echo 

installer-resources/configure-tomcat.sh
if [ -z "$logfile" ]; then
    pushd `dirname $0` > /dev/null
    mydir=`pwd`
    popd > /dev/null

    logfile="$mydir/configure-tomcat.log"
	
	echo "* logfile: $logfile"
fi

echo "* Configuring TOMCAT SERVICE" | tee -a $logfile
sed "s|{YOUR_JAVA_HOME}|$JHF|g" installer-resources/tomcat.service.ini > /etc/systemd/system/tomcat.service

system_memory=0
system_memory=$(grep MemTotal /proc/meminfo | awk '{print $2}')

if ((system_memory < 655360)); then
    newmem=256
    echo "* WARNING"                                                               | tee -a $logfile
    echo "** Your system has very low memory: $system_memory kb"                   | tee -a $logfile
    echo "** The configuration will try to set the java virtual machine to use "   | tee -a $logfile
    echo "** very little memory ($newmem MB). You can change this setting later"   | tee -a $logfile
    echo "** in the file /etc/systemd/system/tomcat.service"                       | tee -a $logfile
    
    sed -i "s|Xms512M|Xms""$newmem""M|g" /etc/systemd/system/tomcat.service
    sed -i "s|Xmx1024M|Xmx""$newmem""M|g" /etc/systemd/system/tomcat.service
elif ((system_memory < 1310720)); then
    newmem=384
    echo "* WARNING"                                                               | tee -a $logfile
    echo "** Your system has low memory: $system_memory kb"                        | tee -a $logfile
    echo "** The configuration will try to set the java virtual machine to use "   | tee -a $logfile
    echo "** little memory ($newmem MB). You can change this setting later"        | tee -a $logfile
    echo "** in the file /etc/systemd/system/tomcat.service"                       | tee -a $logfile
    
    sed -i "s|Xms512M|Xms""$newmem""M|g" /etc/systemd/system/tomcat.service
    sed -i "s|Xmx1024M|Xmx""$newmem""M|g" /etc/systemd/system/tomcat.service
else
	echo "** Server memory: $system_memory kb" | tee -a $logfile
	echo "** java VM memory: Xms512M Xmx1024M"
fi

cp installer-resources/tomcat-context.xml /opt/tomcat/conf/context.xml
cp installer-resources/tomcat-users.xml /opt/tomcat/conf/
cp installer-resources/tomcat-managers-context.xml /opt/tomcat/webapps/manager/META-INF/context.xml
cp installer-resources/tomcat-managers-context.xml /opt/tomcat/webapps/host-manager/META-INF/context.xml

echo "* adding user $SUDO_USER to group tomcat" | tee -a $logfile
if [ -z "$SUDO_USER" ]; then
	echo "** WARNING: Current user could not be detected and added to group. Are you executing with root instead of sudo?" | tee -a $logfile
else
    usermod -a -G tomcat $SUDO_USER | tee -a $logfile
fi


systemctl daemon-reload
systemctl start tomcat
systemctl enable tomcat

ufw allow 8080

echo "* TOMCAT STATUS:" | tee -a $logfile
systemctl status tomcat | grep Active: | tee -a $logfile
echo
echo "******************************************************************************"

migration-from-hosted/migrate-from-hosted.sh
#This file is part of the jobsworth-installer program
#https://github.com/frank-orellana/jobsworth-installer
#Copyright (C) 2017  Franklin Orellana
#This program is free software; you can redistribute it and/or modify
#it under the terms of the GNU General Public License as published by
#the Free Software Foundation; either version 3 of the License, or
#(at your option) any later version.
#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.
#You should have received a copy of the GNU General Public License
#along with this program; if not, write to the Free Software Foundation,
#Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
#WARNING:
#-This scrpit will overwrite info you may have on the jobsworth

backup=$1

if [ -z "$backup" ]
  then
    echo You must specify a backup file:
	echo migrate-from-hosted.sh my_backup_file.sql
	exit
fi

if [ ! -f "$backup" ]; then
    echo Backup file not found: \"$backup\"
	echo You must spscify a valid backup file. Try entering the full path
	exit
fi

echo Starting migration with file: $backup

echo "******************************************************************************"
echo Creating clockingit database to load backup
echo
echo Please enter your MySql root password
mysql -u root -p < schema.sql
echo
echo Schema creation finalized. Please check if there has been any errors
echo "******************************************************************************"
read -n 1 -s -p "Press ENTER to continue"
echo
echo "******************************************************************************"
echo Loading backup into clockingit database
echo
echo Please enter your MySql root password
mysql -u root -p clockingit < $backup
echo
echo Data loaded. Please check if there has been any errors
echo "******************************************************************************"
read -n 1 -s -p "Press ENTER to continue"
echo
echo "******************************************************************************"
echo Migrating data from database clockingit to jobsworth
echo
echo Please enter your MySql root password
mysql -u root -p < data-migration.sql
echo
echo Schema creation finalized. Please check if there has been any
echo "******************************************************************************"
read -n 1 -s -p "Press ENTER to continue"
echo
echo "******************************************************************************"
echo Data migration finalized
echo "******************************************************************************"


install-jobsworth.sh
#!/bin/bash

#This file is part of the jobsworth-installer program
#https://github.com/frank-orellana/jobsworth-installer
#Copyright (C) 2017  Franklin Orellana
#This program is free software; you can redistribute it and/or modify
#it under the terms of the GNU General Public License as published by
#the Free Software Foundation; either version 3 of the License, or
#(at your option) any later version.
#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.
#You should have received a copy of the GNU General Public License
#along with this program; if not, write to the Free Software Foundation,
#Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
#WARNING:
#-This scrpit will overwrite info you may have on the jobsworth

echo "Starting installation..." 

pushd `dirname $0` > /dev/null
mydir=`pwd`
popd > /dev/null

logfile="$mydir/install.log"
echo "* logfile: $logfile"

if (( $EUID != 0 )); then
    echo 'ERROR: You dont have sudo rights. Execute the file like this: sudo install-jobsworth-from-scratch.sh' | tee $logfile
    exit
else
	echo '* Starting process...' | tee $logfile
fi

if [ -z "$NO_UPD_REP" ]; then
	echo "* Updating repositories..." | tee -a $logfile
	apt-get -qq update
fi

if [ -z "$NO_DB_SERVER" ]; then
	echo "* Installing mysql (MariaDB)" | tee -a $logfile

	echo "**   If it is not already installed it will ask you for a new password for "
	echo "**   the ROOT user. DON'T FORGET THAT PASSWORD!!!"
	echo "**   It will be asked later to create the default database"
	echo "**   If no password is requested the default password is blank"
	echo "** Downloading..."
	apt-get -qqy install mariadb-server | tee -a $logfile
else
	echo "** DB Server installation skipped" | tee -a $logfile
fi

if [ -z "$NO_DB" ]; then
	echo "** Creating Jobsworth database."
	echo "** Executing create-database.sql script... Enter mysql (MariaDB) root password:" | tee -a $logfile
	mysql -u root -p < installer-resources/create-database.sql
	echo
	echo "******************************************************************************"
	echo "** Jobsworth database created. To access it you can type 'mysql -u jobsworth -p'" | tee -a $logfile
	echo "******************************************************************************"
else
	echo "** DB creation skipped" | tee -a $logfile
fi

echo

if [ -z "$NO_JDK" ]; then
	echo "* Installing JAVA" | tee -a $logfile

	JHF=$(update-java-alternatives -l | grep -o '/.*jvm/java-1.[89].*$')
	if [ -z "$JHF" ]; then
		java_available=$(apt-cache search openjdk-8-jdk)
		
		if [ -z "$java_available" ]; then
			echo "** JAVA package not available. Adding source http://http.debian.net/debian jessie-backports main" | tee -a $logfile
			echo deb http://http.debian.net/debian jessie-backports main >> /etc/apt/sources.list
			echo "** Updating repositories..." | tee -a $logfile
			apt-get -qq update
			echo "** Repositories updated" | tee -a $logfile
			
			echo "** Installing java openjdk-8-jdk..." | tee -a $logfile
			apt-get -qqy install -t jessie-backports  openjdk-8-jre-headless ca-certificates-java openjdk-8-jdk
		else
			echo "** openjdk-8-jdk is available to install"
			echo "** Installing java openjdk-8-jdk..." | tee -a $logfile
			apt-get -qqy install openjdk-8-jre-headless ca-certificates-java openjdk-8-jdk  | tee -a $logfile
		fi
		
		echo "** openjdk-8-jdk installed..." | tee -a $logfile

		JHF=$(update-java-alternatives -l | grep -o '/.*jvm/java-1.[89].*$')
		if [ -z "$JHF" ]; then 
			echo "* ERROR: JAVA_HOME Not Found. Exiting" | tee -a $logfile
			update-java-alternatives -l  | tee -a $logfile
			exit 
		else 
			echo "* JAVA_HOME FOUND: $JHF" | tee -a $logfile
			JHF=$JHF/jre
		fi
	else
		echo "** JAVA JDK 8 or 9 is already installed in $JHF" | tee -a $logfile
		JHF=$JHF/jre
	fi
else
	echo "** JDK Installation skipped" | tee -a $logfile
fi

if [ -z "$NO_TOMCAT" ]; then
	groupadd tomcat
	useradd -s /bin/false -g tomcat -d /opt/tomcat tomcat | tee -a $logfile

	echo "* Downloading TOMCAT..." | tee -a $logfile
	cd $mydir
	wget --continue -nv http://archive.apache.org/dist/tomcat/tomcat-8/v8.5.15/bin/apache-tomcat-8.5.15.tar.gz | tee -a $logfile

	echo "* TOMCAT dowloaded" | tee -a $logfile

	mkdir -p /opt/tomcat | tee -a $logfile
	tar -xzf apache-tomcat-8*tar.gz -C /opt/tomcat --strip-components=1

	echo "* TOMCAT extracted" | tee -a $logfile

	echo "* Configuring TOMCAT..." | tee -a $logfile
	cd /opt/tomcat
	chgrp -R tomcat /opt/tomcat
	chmod -R g+r conf
	chmod g+x conf
	chown -R tomcat webapps/ work/ temp/ logs/
	chmod 775 -R conf/ logs/ webapps/

	cd conf
	find . -type f -exec chmod 664 -- {} +

	cd ../webapps
	find . -type f -exec chmod 664 -- {} +

	cd $mydir

	chmod +x installer-resources/configure-tomcat.sh | tee -a $logfile
	. ./installer-resources/configure-tomcat.sh


	echo "* TOMCAT has been installed. Please check its installation visiting http://localhost:8008 OR HTTP://<THIS_COMPUTER_IP>:8080 before continuing"
	read -n 1 -s -p "Press ENTER to continue"
else
	echo "** TOMCAT Installation skipped" | tee -a $logfile
fi

echo
echo "* Stopping TOMCAT"
systemctl stop tomcat

echo "* Downloading and Installing Jobsworth" | tee -a $logfile
cd /var
mkdir -p jobsworth/assets | tee -a $logfile

chown -R tomcat jobsworth
chgrp -R tomcat jobsworth
chmod -R 775 jobsworth

cd /opt/tomcat
rm -rf webapps/ROOT


echo "* Downloading WAR" | tee -a $logfile
cd $mydir
wget --continue -nv https://github.com/ari/jobsworth/releases/download/5.0b2/ROOT.war  | tee -a $logfile
echo "* WAR Downloaded" | tee -a $logfile
cp ROOT.war /opt/tomcat/webapps
cd /opt/tomcat/webapps
chown tomcat ROOT.war
chgrp tomcat ROOT.war

systemctl start tomcat
echo "* TOMCAT STATUS:" | tee -a $logfile
systemctl status tomcat | grep Active: | tee -a $logfile

chmod 664 /opt/tomcat/logs/*
echo
echo "******************************************************************************"
echo "******************************************************************************"
echo "JOBSWORTH HAS BEEN INSTALLED"  | tee -a $logfile
echo "THE INSTALLATION IS COMPLETE TYPE HTTP://LOCALHOST:8080 in the browser from "
echo "this computer OR HTTP://<THIS_COMPUTER_IP>:8080 FROM ANY OTHER"
echo "******************************************************************************"
echo "******************************************************************************"
echo 